Mathlib/AlgebraicTopology/SimplicialSet/Coskeletal.lean

```diff
@@ -0,0 +1,255 @@
+            rfl
+      rw [← StructuredArrow.w β₁, FunctorToTypes.map_comp_apply, this, ← s.w β₁]
+      dsimp
+
+lemma fac_aux₃ {n : ℕ}
+    (s : Cone (proj (op [n]) (Truncated.inclusion 2).op ⋙ (Truncated.inclusion 2).op ⋙ X))
+    (x : s.pt) (φ : [1] ⟶ [n]) :
+    X.map φ.op (lift s x) = s.π.app (strArrowMk₂ φ (by omega)) x := by
+  obtain ⟨i, j, hij, rfl⟩ : ∃ i j hij, φ = mkOfLe i j hij :=
+    ⟨φ.toOrderHom 0, φ.toOrderHom 1, φ.toOrderHom.monotone (by simp),
+      Hom.ext_one_left _ _ rfl rfl⟩
+  exact fac_aux₂ _ _ _ _ _ _ (by omega)
+
+end isPointwiseRightKanExtensionAt
+
+open Truncated
+
+open isPointwiseRightKanExtensionAt in
+/-- A strict Segal simplicial set is 2-coskeletal. -/
+noncomputable def isPointwiseRightKanExtensionAt (n : ℕ) :
+    (rightExtensionInclusion X 2).IsPointwiseRightKanExtensionAt ⟨[n]⟩ where
+  lift s x := lift (X := X) s x
+  fac s j := by
+    ext x
+    obtain ⟨⟨i, hi⟩, ⟨f⟩, rfl⟩ := j.mk_surjective
+    obtain ⟨i, rfl⟩ : ∃ j, SimplexCategory.mk j = i := ⟨_, i.mk_len⟩
+    change [i] ⟶ [n] at f
+    dsimp at hi ⊢
+    apply StrictSegal.spineInjective
+    dsimp
+    ext k
+    · dsimp only [spineEquiv, Equiv.coe_fn_mk]
+      erw [spine_map_vertex]
+      rw [spine_spineToSimplex, spine_vertex]
+      let α : strArrowMk₂ f hi ⟶ strArrowMk₂ ([0].const [n] (f.toOrderHom k)) (by omega) :=
+        StructuredArrow.homMk (([0].const _ (by exact k)).op) (by simp; rfl)
+      exact congr_fun (s.w α).symm x
+    · dsimp only [spineEquiv, Equiv.coe_fn_mk, spine_arrow]
+      rw [← FunctorToTypes.map_comp_apply]
+      let α : strArrowMk₂ f hi ⟶ strArrowMk₂ (mkOfSucc k ≫ f) (by omega) :=
+        StructuredArrow.homMk (mkOfSucc k).op (by simp; rfl)
+      exact (isPointwiseRightKanExtensionAt.fac_aux₃ _ _ _ _).trans (congr_fun (s.w α).symm x)
+  uniq s m hm := by
+    ext x
+    apply StrictSegal.spineInjective (X := X)
+    dsimp [spineEquiv]
+    rw [StrictSegal.spine_spineToSimplex]
+    ext i
+    · exact congr_fun (hm (StructuredArrow.mk (Y := op [0]₂) ([0].const [n] i).op)) x
+    · exact congr_fun (hm (.mk (Y := op [1]₂) (.op (mkOfLe _ _ (Fin.castSucc_le_succ i))))) x
+
+/-- Since `StrictSegal.isPointwiseRightKanExtensionAt` proves that the appropriate
+cones are limit cones, `rightExtensionInclusion X 2` is a pointwise right Kan extension.-/
+noncomputable def isPointwiseRightKanExtension :
+    (rightExtensionInclusion X 2).IsPointwiseRightKanExtension :=
+  fun Δ => isPointwiseRightKanExtensionAt X Δ.unop.len
+
+theorem isRightKanExtension :
+    X.IsRightKanExtension (𝟙 ((inclusion 2).op ⋙ X)) :=
+  RightExtension.IsPointwiseRightKanExtension.isRightKanExtension
+    (isPointwiseRightKanExtension X)
+
+/-- When `X` is `StrictSegal`, `X` is 2-coskeletal. -/
+instance isCoskeletal : SimplicialObject.IsCoskeletal X 2 where
+  isRightKanExtension := isRightKanExtension X
+
+end StrictSegal
+
+end
+
+end SSet
+
+namespace CategoryTheory
+
+namespace Nerve
+
+open SSet
+
+example (C : Type u) [Category.{v} C] :
```

Comment:
I don't understand the purpose of this example. If it is documentation, then no one will see it outside of a doc comment. If it is to make sure instance synthesis succeeds, then it should be a test importing more than this single file.

Reply:
Hi @mattrobball. I don't understand what you mean by "...then it should be a test importing more than this single file." Could you explain?

That instance gets used implicitly in the final term `CategoryTheory.Nerve.cosk₂Iso` in the file. But as you guessed, instance synthesis succeeds. That term is still definable if this example is deleted.

On the other hand, to a human reader, this example and the consequent isomorphism is the whole point of this file (the main theorem). If it's consistent with style guidelines, I'd like to keep this as an explicit term so that when someone searches for this result later they'll be able to find it, without having to figure out that this is a special case of the more abstract theorem proven here (which is less "well-known"). I'd added an explicit mention to this fact in the comments at the top of the file.

Does this work for you, or would you like me to make further changes?

Reply:
I think it's fine like this.

Reply:
I'm not a big fan of `example`s in the code base because they tend to add to the imports and overall build time. They also confound environmental linters since nothing added to the environment.  

If keeping it as active code for documentation is important, I think declaring it as instance would cause less problems. But including the example as a code block in the comment itself would probably reach more eyes than doing this. 

Sometimes these examples are used to assure that instance synthesis doesn't fail under future chances. I don't think that is the intention here but I could be wrong. If it was, then one would want to assure that instance synthesis keeps working thoughout mathlib, ie nothing downstream breaks it. Having the example here does not assure this.

Reply:
Those are actually good points. Maybe we should codify this more in the contributor guidelines?

Reply:
One could also make it a `def` to avoid adding an instance but also making it visible to any environmental searching.  

> Those are actually good points. Maybe we should codify this more in the contributor guidelines?

I agree.

Reply:
Otoh, in this specific case we aren't adding extra imports, right?
I think there is value to having this as active code, even in this file.

It can also be added in the docstring, but since it is already mentioned in the module docstring I think we're good.

Let's not block this PR on this issue. We can always change/improve later.

bors merge

Reply:
We should make it an honest declaration then. It would come at no cost and some benefit.